### PR TITLE
feat(ecs-patterns): add listener port as a property for network/application load balanced services

### LINKED ISSUE #4793: NetworkLoadBalancedFargateService to support listener other than tcp80
At this moment `NetworkLoadBalancedFargateService` is extending `NetworkLoadBalancedServiceBase`, which only support listener on tcp80.

https://github.com/aws/aws-cdk/blob/d7654e714dc940565d76925dcbbb976cef6ddae4/packages/%40aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts#L248

Please support different port other than 80.

### Use Case

For some TCP application running on Fargate behind NLB, it's very common to listen on port other than 80.

### Proposed Solution

Add a `listenerPort` property for `NetworkLoadBalancedServiceBaseProps`.

https://github.com/aws/aws-cdk/blob/d7654e714dc940565d76925dcbbb976cef6ddae4/packages/%40aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts#L12-L110

### Other

* [v] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by pahud**
Allow users to specify the listener port as input to application/network load balanced fargate/ecs services. 
- If a port is specified, both the target group and listener port will be overridden.
- If a port is not specified, the target group port will be defaulted to 80 and the listener port will default to 80 if the protocol is HTTP or 443 if the protocol is HTTPS.

Fixes: #4793 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by karlpatr** [comment]
`assignPublicIp enabled should allow all ipv4 traffic`

I disagree with this statement. I have multiple AWS services that are meant to be accessed only from inside a corporate VPN.  It has public IPs because the traffic comes from outside the AWS account, but the access to the service behind the NLB is still restricted with a prefix list specifying the corporate IP block.

I don't use the patterns helper functions, but if this change is accepted the configuration I support will be impossible to build because the assumption made is not globally true.

I strongly encourage making any change that automatically reduces the security of a system for convenience explicitly opt-in.

**Comment by pkandasamy91** [comment]
> `assignPublicIp enabled should allow all ipv4 traffic`
> 
> I disagree with this statement. I have multiple AWS services that are meant to be accessed only from inside a corporate VPN. It has public IPs because the traffic comes from outside the AWS account, but the access to the service behind the NLB is still restricted with a prefix list specifying the corporate IP block.
> 
> I don't use the patterns helper functions, but if this change is accepted the configuration I support will be impossible to build because the assumption made is not globally true.
> 
> I strongly encourage making any change that automatically reduces the security of a system for convenience explicitly opt-in.

I'm in agreement here, I don't think the use-case you handle can be described as the general use-case. I would rather not reduce the security levels of the construct and would leave it up to the user to do it after the fact if they prefer. 

Since your PR is it's around allowing different listener ports (nlb/alb), please reduce the scope of this PR to only allow a listener port to be input to the construct and make any additional changes in a separate PR.

File: packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts [hunk 1]
```diff
@@ -135,5 +140,11 @@ export class NetworkLoadBalancedFargateService extends NetworkLoadBalancedServic
       cloudMapOptions: props.cloudMapOptions,
     });
     this.addServiceAsTarget(this.service);
+
+    // all public-facing NLB fronted fargate service with assignPublicIp should allow from any IPv4 traffic
+    const publicNLB = props.publicLoadBalancer !== undefined ? props.publicLoadBalancer : true;
+    if (publicNLB && props.assignPublicIp) {
+      this.service.connections.connections.allowFromAnyIpv4(Port.tcp(containerPort));
+    }
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts]: Please submit this change as a separate PR.

File: packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts [hunk 2]
```diff
@@ -123,6 +124,10 @@ export class NetworkLoadBalancedFargateService extends NetworkLoadBalancedServic
       throw new Error('You must specify one of: taskDefinition or image');
     }
 
+    // determine the ingress container port
+    const containerPort = props.taskImageOptions ? props.taskImageOptions.containerPort || 80 :
+      props.taskDefinition ? props.taskDefinition.defaultContainer!.ingressPort : 80;
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts]: Please submit this change as a separate PR.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts [hunk 3]
```diff
@@ -86,6 +86,15 @@ export interface NetworkLoadBalancedServiceBaseProps {
    */
   readonly loadBalancer?: NetworkLoadBalancer;
 
+  /**
+   * Listener port of the network load balancer that will serve traffic to the service.
+   *
+   * [disable-awslint:ref-via-interface]
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: Not sure why you're disabling the linter here. Please remove this line.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts [hunk 4]
```diff
@@ -86,6 +86,15 @@ export interface NetworkLoadBalancedServiceBaseProps {
...
+   *
+   * @default - 80
+   */
+  readonly listenerPort?: number;
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: Is there any reason why this wouldn't apply to the ApplicationLoadBalancedServiceBase as well?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts (hunk 3 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: removed

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts (hunk 4 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: support ApplicationLoadBalancedServiceBase now

File: packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts (hunk 2 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts]: OK. removed.

File: packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts (hunk 1 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-load-balanced-fargate-service.ts]: OK removed.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 5]
```diff
@@ -108,6 +108,13 @@ export interface ApplicationLoadBalancedServiceBaseProps {
    */
   readonly loadBalancer?: ApplicationLoadBalancer;
 
+  /**
+   * Listener port of the application load balancer that will serve traffic to the service.
+   *
+   * @default - defaults to 80 if not specified
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Please provide the default as only the value. 

`@default 80`

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 6]
```diff
@@ -273,17 +280,21 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
...
-    const targetProps = {
-      port: 80
-    };
-
     if (props.certificate !== undefined && props.protocol !== undefined && props.protocol !== ApplicationProtocol.HTTPS) {
       throw new Error('The HTTPS protocol must be used when a certificate is given');
     }
     const protocol = props.protocol !== undefined ? props.protocol : (props.certificate ? ApplicationProtocol.HTTPS : ApplicationProtocol.HTTP);
 
+    const listenerPort = props.listenerPort !== undefined ? props.listenerPort : protocol === ApplicationProtocol.HTTPS ? 443 : 80;
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: I like what you've done here, it might be worth opening an issue to track this.

The issue is this is a breaking change. If a user has a production service running with HTTPS but using the previous construct, they would currently be running on 80, not 443. 

`const listenerPort = props.listenerPort !== undefined ? props.listenerPort : 80;`

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts [hunk 7]
```diff
@@ -86,6 +86,13 @@ export interface NetworkLoadBalancedServiceBaseProps {
    */
   readonly loadBalancer?: NetworkLoadBalancer;
 
+  /**
+   * Listener port of the network load balancer that will serve traffic to the service.
+   *
+   * @default - defaults to 80 if not specified
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: Please provide the default as only the value.

@default 80

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 6 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: OK created https://github.com/aws/aws-cdk/issues/4861

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts (hunk 7 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: fixed

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 5 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: fixed

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 6 above)
Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: You should update the check to be `const listenerPort = props.listenerPort !== undefined ? props.listenerPort : 80;` This module is stable and we cannot accept a breaking change.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts [hunk 8]
```diff
@@ -86,6 +86,13 @@ export interface NetworkLoadBalancedServiceBaseProps {
    */
   readonly loadBalancer?: NetworkLoadBalancer;
 
+  /**
+   * Listener port of the network load balancer that will serve traffic to the service.
+   *
+   * @default - 80
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: please remove the `-` the default should be specified as `@default 80`

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 6 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: OK. I got it reverted.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts (hunk 8 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: fixed

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 9]
```diff
@@ -108,6 +108,13 @@ export interface ApplicationLoadBalancedServiceBaseProps {
    */
   readonly loadBalancer?: ApplicationLoadBalancer;
 
+  /**
+   * Listener port of the application load balancer that will serve traffic to the service.
+   *
+   * @default - 80
```
[the remark this answers is not in the document]
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: fixed

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 6 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: I probably need this otherwise I will break a [test](https://github.com/aws/aws-cdk/pull/4825#issuecomment-549938000).

That is, when no `listenerPort` specified, pass it as `undefined` to avoid the breaking.

```js
 const listenerPort = props.listenerPort !== undefined ? props.listenerPort : undefined;

    const targetProps = {
      port: listenerPort,
      protocol
    };

    this.listener = this.loadBalancer.addListener('PublicListener', {
      protocol,
      port: listenerPort,
      open: true
    });
    this.targetGroup = this.listener.addTargets('ECS', targetProps);
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 10]
```diff
@@ -273,17 +280,21 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
...
+    const listenerPort = props.listenerPort;
+
+    const targetProps = {
+      port: listenerPort,
+      protocol
+    };
+
     this.listener = this.loadBalancer.addListener('PublicListener', {
       protocol,
+      port: listenerPort,
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: I believe the test is failing because you're specifying the port here using the same port that is specified as the listener port. If you remove this line and add the default, the failing test should pass.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 11]
```diff
@@ -273,17 +280,21 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
...
-    const targetProps = {
-      port: 80
-    };
-
     if (props.certificate !== undefined && props.protocol !== undefined && props.protocol !== ApplicationProtocol.HTTPS) {
       throw new Error('The HTTPS protocol must be used when a certificate is given');
     }
     const protocol = props.protocol !== undefined ? props.protocol : (props.certificate ? ApplicationProtocol.HTTPS : ApplicationProtocol.HTTP);
 
+    const listenerPort = props.listenerPort;
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: You should leave the default that was originally defined here port 80.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 12]
```diff
@@ -273,17 +280,21 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
 
     this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer : new ApplicationLoadBalancer(this, 'LB', lbProps);
 
-    const targetProps = {
-      port: 80
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: What you could do is leave this as is, and just replace `port: 80` with `port: props.listenerPort !== undefined ? props.listenerPort : 80;`

The rest of the changes below are not necessary to support listener ports.

Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Thanks for the insight. I will continue work on it tomorrow. Great to have learned a lot from it.

And will add the integ and unit testing as well.

**Comment by pkandasamy91** [comment]
Can you also add a unit test to test when you specify a different port?

**Comment by pahud** [comment]
Hi @pkandasamy91 

All tests passed including some additional test coverage(https://github.com/aws/aws-cdk/pull/4825/commits/73525c70866b23a4c03825c5c2b374a329db095d). Can you kindly review it again?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 12 above)
Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: This has yet to be resolved.

**Comment by pkandasamy91** [comment]
@pahud, thanks for adding the tests, but you have yet to resolve the comments from the previous review.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 13]
```diff
@@ -273,17 +280,19 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
...
-    };
-
     if (props.certificate !== undefined && props.protocol !== undefined && props.protocol !== ApplicationProtocol.HTTPS) {
       throw new Error('The HTTPS protocol must be used when a certificate is given');
     }
     const protocol = props.protocol !== undefined ? props.protocol : (props.certificate ? ApplicationProtocol.HTTPS : ApplicationProtocol.HTTP);
 
+    const targetProps = {
+      port: props.listenerPort !== undefined ? props.listenerPort : 80,
```
Comment (pahud) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: modified as suggested

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 14]
```diff
@@ -273,17 +280,19 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
...
+      protocol
+    };
+
     this.listener = this.loadBalancer.addListener('PublicListener', {
       protocol,
+      port: props.listenerPort,
```
Comment (pahud) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: I need to keep this line otherwise the test will fail

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 12 above)
Comment (pahud) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: I have made some fix per requested and all tests passed. Please review it again. Thanks.

**Comment by pahud** [comment]
@pkandasamy91 

Just fixed the lines as suggested and all tests passed.
